### PR TITLE
Support for WebSocket handshake

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves item 2 in https://github.com/microsoft/CCF/issues/629. 

This PR adds support for the WebSocket handshake. 

Typically, the (websocket) client sends a handshake request over HTTP, specifying the upgrade protocol and a random key, e.g.:
```
GET / HTTP/1.1
Upgrade: websocket
Connection: Upgrade
Host: 127.237.61.243:46920
Origin: http://127.237.61.243:46920
Sec-WebSocket-Key: tuCthd5nFUGth+AnhtfWCQ==
Sec-WebSocket-Version: 13
```

To complete the upgrade, the server should reply with:
```
HTTP/1.1 101 Switching Protocols
Connection: Upgrade
Sec-WebSocket-Accept: +y6XCTIDNuuThvP3OSUoiAUA+UY=
Upgrade: websocket
```

The `Sec-Websocket-Accept` string is constructed by 1) concatenating the client's key with a well-known string, 2) SHA1 it, 3) Base 64 encode the result. 

- Added a websocket client in `infra.clients.py` which only tests the handshake. 
- Added a HTTP Response builder in `httpbuilder.h` 
- Also changed the HTTP Signature parser to use static methods so that the interface is nicer.

Still to do: 
- Process the WebSockets frames once the endpoint has been upgraded.

### CODE REVIEW

**Comment by cimetrics** [comment]
images

File: tests/infra/clients.py [hunk 1]
```diff
@@ -423,6 +468,8 @@ def __init__(self, *args, **kwargs):
         if os.getenv("HTTP"):
             if os.getenv("CURL_CLIENT"):
                 self.client_impl = CurlClient(*args, **kwargs)
+            elif os.getenv("WEBSOCKETS_CLIENT"):
+                self.client_impl = WSClient(*args, **kwargs)
```
Comment (achamayou) [inline tests/infra/clients.py]: I don't think this should be a separate client? Presumably we just have .upgrade() on our existing requests-based client?

File: tests/infra/clients.py [hunk 2]
```diff
@@ -413,6 +414,50 @@ def disconnect(self):
...
+            sslopt={"certfile": self.cert, "keyfile": self.key, "ca_certs": self.ca},
+        )
+        # TODO: Support sending data over websocket
+        # ws.send(request.to_json())
+        # res = ws.recv()
+        # ws.close()
+        # self.stream.update(rep.content)
+        return request.id
+
+    def signed_request(self, request):
```
Comment (achamayou) [inline tests/infra/clients.py]: That's going to need the outer flatbuffer frame to be setup before it can be done.

**Comment by cimetrics** [comment]
images

File: tests/infra/clients.py (hunk 1 above)
Comment (jumaffre) [reply tests/infra/clients.py]: Hmmm, not sure about this one. As I see it, clients would _not_ use a http client for some requests, then upgrade to ws. Instead, I think they would connect to `wss://node0.ccf.io:55555` from the start and the endpoint would be upgraded to ws straight away. 

Also, AFAICS Python's `requests` does not support upgrading to ws so we would need to create a new endpoint in the enclave on upgrade (i.e. when `create_connection()` is called).

File: tests/infra/clients.py (hunk 2 above)
Comment (jumaffre) [reply tests/infra/clients.py]: Or, if we support plain JSON-RPC over ws, we can use the old signature scheme.

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: tests/infra/clients.py (hunk 2 above)
Comment (achamayou) [reply tests/infra/clients.py]: We won't do that, we will only support FlatBuffers over WS, unless there's a clear use-case for something else.

File: tests/infra/clients.py (hunk 1 above)
Comment (achamayou) [reply tests/infra/clients.py]: Ah fair enough then.

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Pull request merged**